### ISSUE TITLE
cdc emit row changed event to downstream kafka may block a long time.

### ISSUE THREAD

**Issue by 3AceShowHand** [pingcap/tiflow#4987]
### What did you do?

1. create a changefeed with kafka sink
2. stop the changfeed
3. prepare `go-tpc` workload
4. resume the changefeed
5. run `go-tpc` workload

periodically use `kill -s STOP` to pause the Kafka process for around 40 ~ 50s, each a few minutes.

### What did you expect to see?

The whole CDC works properly.

### What did you see instead?

Processor gets blocked for a long time.

When the problem happen, kafka process is healthy, resumed by `kill -s CONT`

```
[2022/03/22 14:00:30.348 +08:00] [WARN] [client.go:263] ["etcd client outCh blocking too long, the etcdWorker may be stuck"] [duration=23.000155792s] [role=processor]
[2022/03/22 14:00:31.348 +08:00] [WARN] [client.go:263] ["etcd client outCh blocking too long, the etcdWorker may be stuck"] [duration=23.9999115s] [role=processor]
[2022/03/22 14:00:32.348 +08:00] [WARN] [client.go:263] ["etcd client outCh blocking too long, the etcdWorker may be stuck"] [duration=24.9999985s] [role=processor]
[2022/03/22 14:00:33.349 +08:00] [WARN] [client.go:263] ["etcd client outCh blocking too long, the etcdWorker may be stuck"] [duration=26.000115375s] [role=processor]
[2022/03/22 14:00:34.349 +08:00] [WARN] [client.go:263] ["etcd client outCh blocking too long, the etcdWorker may be stuck"] [duration=27.000068583s] [role=processor]
[2022/03/22 14:00:35.349 +08:00] [WARN] [client.go:263] ["etcd client outCh blocking too long, the etcdWorker may be stuck"] [duration=28.000139292s] [role=processor]
[2022/03/22 14:00:36.348 +08:00] [WARN] [client.go:263] ["etcd client outCh blocking too long, the etcdWorker may be stuck"] [duration=28.999832417s] [role=processor]
[2022/03/22 14:00:37.349 +08:00] [WARN] [client.go:263] ["etcd client outCh blocking too long, the etcdWorker may be stuck"] [duration=29.99986875s] [role=processor]
[2022/03/22 14:00:38.349 +08:00] [WARN] [client.go:263] ["etcd client outCh blocking too long, the etcdWorker may be stuck"] [duration=30.999821875s] [role=processor]
```
### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
master
```

Upstream TiKV version (execute `tikv-server --version`):

```console
master
```

TiCDC version (execute `cdc version`):

```console
master
```

[etcd_worker.log](https://github.com/pingcap/tiflow/files/8322983/etcd_worker.log)
[goroutine2.txt](https://github.com/pingcap/tiflow/files/8322990/goroutine2.txt)

**Comment by 3AceShowHand** [comment]
![CzKJCctpFH](https://user-images.githubusercontent.com/7138436/159449502-fc409d48-a9c1-4f6f-890c-071ce0f535e1.png)

It looks that `EmitRowChangedEvent` blocked a long time.

**Comment by 3AceShowHand** [comment]
there is also dead lock among buffer sink.

**Comment by liuzix** [comment]
A summary of the series of events:
1. Sarama throws an error through `asyncSend`.
2. The error causes the `run` goroutine of `mqFlushWorker` to exit.
3. Without a consumer, the producer to the `msgChan` in `mqFlushWorker` is blocked, which blocks `EmitRowChangedEvent` in `mqSink`.
4. The goroutine blocked by `mqFlushWorker` holds `bufferMu` in `buffersink`.
5. `bufferSink` is unable to be written to, blocking the pipeline channels in all tables.
6. The pipelines cannot be canceled by `(*processor).Close`.
7. Since `(*processor).Close` is called in main goroutine of the EtcdWorker, it blocks all operations to all changefeeds.

**Comment by 3AceShowHand** [comment]
closed by #4996
